network_simulator/flow.py

```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
         return str(self.__dict__)
 
-# TODO(team): Flow-related Event subclasses
-# TODO(team): Flow subclasses for at least two TCP algorithms.
+
+
+class InitiateFlowEvent(Event):
+    def __init__(self, flow=None, packet_ids=[]):
+        """
+        An one-time Event called to set up the Flow at a given timestamp.
+
+        :ivar Flow flow: flow of this event.
+        :ivar list packet_ids: a list of integers of packet IDs.
+        :ivar list packets_to_send: list of data packets to send. This is
+        default to an empty list. This list will be populated later.
+        """
+        self.flow = flow
+        self.packet_ids = packet_ids
+        self.packets_to_send = []
+
+    def create_data_packet(self, packet_id, curr_time):
+        """
+        Create a DataPacket based on the current flow.
+
+        :param int packet_id: corresponding packet ID.
+        :param float curr_time: simulation time of this function call.
+        :return DataPacket
+        """
+        packet = DataPacket(packet_id, self.flow.flow_id,
+                            self.flow.source_addr, self.flow.dest_addr,
+                            curr_time)
+        return packet
+
+    def run(self, main_event_loop, statistics):
+        """
+        Set up a list of data packets to initiate a flow.
+
+        :param MainEventLoop main_event_loop: main loop to retrieve the
+        simulation time of function call.
+        :param Statistics statistics: the Statistics to update.
+        """
+        for curr_packet in self.packet_ids:
+            if not self.flow.packet_id_exceeds_data(curr_packet):
+                new_data_packet = self.create_data_packet(
+                    curr_packet, main_event_loop.global_clock_sec)
+                self.packets_to_send.append(new_data_packet)
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedule a FlowSendPacketsEvent immediately, with the list of packets
+        to send.
+
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        flow_send_event = FlowSendPacketsEvent(self.flow,
+                                               self.packets_to_send)
+        main_event_loop.schedule_event_with_delay(flow_send_event, 0.0)
+
+
+
+class FlowSendPacketsEvent(Event):
+    def __init__(self, flow, packets_to_send):
+        """
+        Send data packet from a flow.
+
+        :ivar Flow flow: flow of this event.
+        :ivar list packets_to_send: list of data packets to send.
+        """
+        self.flow = flow
+        self.packets_to_send = packets_to_send
+
+    def run(self, main_event_loop, statistics):
+        """
+        Check that the size of packets combined is smaller than
+        the window size.
+        :param MainEventLoop main_event_loop: main loop.
+        :param Statistics statistics: the Statistics to update.
+        """
+        if len(set(self.flow.packets_in_transit) \
```

Comment:
You will have to do set(...).union(set(...)). Python doesn't recognize + with sets means set addition.